File: solutions/c++/problems/[0014_Easy] Longest Common Prefix.cpp
struct TrieNode {
    unordered_map<char, TrieNode *> children;
};

class Trie {
    int minLen = INT_MAX;
    TrieNode *root = new TrieNode();
public:
    Trie() {}

    void add(string word) {
        minLen = min(minLen, (int) word.size());
        auto curr = root;
        for (char c: word) {
            if (!curr->children[c]) curr->children[c] = new TrieNode();
            curr = curr->children[c];
        }
    }

    string getLongestPrefix() {
        string ans = "";
        auto curr = root;
        while (curr->children.size() == 1 && ans.size() < minLen) {
            auto child = curr->children.begin();
            ans += child->first;
            curr = child->second;
        }
        return ans;
    }
};

class Solution {
public:
    string longestCommonPrefix(vector <string> &strs) {
        auto trie = new Trie();
        for (auto word: strs) trie->add(word);
        return trie->getLongestPrefix();
    }
};

File: solutions/java/problems/[0014_Easy] Longest Common Prefix.java
class Solution {
    public String longestCommonPrefix(String[] strs) {
        int n = strs.length;
        StringBuilder ans = new StringBuilder(strs[0]);
        int matching = strs[0].length() - 1;
        for (int i = 1; i < n; i++) {
            String word = strs[i];
            matching = Math.min(matching, word.length() - 1);
            for (int j = 0; j <= matching; j++) {
                if (word.charAt(j) != ans.charAt(j)) {
                    matching = j - 1;
                    break;
                }
            }
        }
        if (matching < 0) return "";
        return ans.substring(0, matching + 1);
    }
}
